src/strands/experimental/tools/executors/executor.py

```diff
@@ -0,0 +1,247 @@
+"""Abstract base class for tool executors.
+
+Tool executors are responsible for determining how tools are executed (e.g., concurrently, sequentially, with custom
+thread pools, etc.).
+"""
+
+import abc
+import logging
+import time
+from concurrent.futures import ThreadPoolExecutor
+from typing import TYPE_CHECKING, Any, Callable, cast
+
+from opentelemetry import trace as trace_api
+
+from ....telemetry.metrics import Trace
+from ....telemetry.tracer import get_tracer
+from ....types.content import Message
+from ....types.tools import ToolChoice, ToolChoiceAuto, ToolConfig, ToolGenerator, ToolResult, ToolUse
+from ...hooks import AfterToolInvocationEvent, BeforeToolInvocationEvent
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ....agent import Agent
+
+logger = logging.getLogger(__name__)
+
+
+class Executor(abc.ABC):
+    """Abstract base class for tool executors."""
+
+    def __init__(self, thread_pool: ThreadPoolExecutor | str | None = "asyncio", skip_tracing: bool = False):
+        """Initialize the executor.
+
+        Args:
+            thread_pool: Thread pool configuration for synchronous tools.
+
+                - "asyncio" (default): Use the asyncio thread pool
+                - ThreadPoolExecutor: Use the provided custom thread pool
+                - None: Run sync tools in the main thread (i.e., blocking)
+
+            skip_tracing: Do not trace call.
+                Useful for direct tool calls where event loop span metadata is not available.
+        """
+        self.thread_pool = thread_pool
+        self.skip_tracing = skip_tracing
+
+    @staticmethod
+    def _trace(stream: Callable) -> Callable:
+        """Decorator that adds tracing and metrics to tool execution.
+
+        Args:
+            stream: The stream method to wrap with tracing.
+
+        Returns:
+            Wrapped stream method with tracing capabilities.
+        """
+
+        async def wrapper(
+            self: "Executor",
+            agent: "Agent",
+            tool_use: ToolUse,
+            tool_results: list[ToolResult],
+            invocation_state: dict[str, Any],
+        ) -> ToolGenerator:
+            """Execute tool with tracing and metrics collection.
+
+            Args:
+                self: The executor instance.
+                agent: The agent for which the tool is being executed.
+                tool_use: Metadata and inputs for the tool to be executed.
+                tool_results: List of tool results from each tool execution.
+                invocation_state: Context for the tool invocation.
+
+            Yields:
+                Tool events with the last being the tool result.
+            """
+            if self.skip_tracing:
```

Comment:
I wonder if we should explicitly have `event_loop_cycle_span`/`event_loop_cycle_trace` passed in as optional, and if they're not set we don't trace?

Reply:
This also seems to skip adding to event_loop_metrics - is that correct?